src/frequenz/sdk/timeseries/_ringbuffer.py

```diff
@@ -225,17 +227,24 @@ def window(
                 if isinstance(self._buffer, list):
                     window += self._buffer[0:end_index]
                 else:
-                    window = np.concatenate((window, self._buffer[0:end_index]))
+                    window = concatenate((window, self._buffer[0:end_index]))
             return window
 
-        def in_gap(gap) -> bool:
-            return gap.contains(start) or gap.contains(end)
-
         # Return a copy if there are none-values in the data
-        if force_copy or any(map(in_gap, self._gaps)):
-            return deepcopy(self._buffer[start_index:end_index])
-
-        return self._buffer[start_index:end_index]
+        if force_copy or any(
+            map(lambda gap: gap.contains(start) or gap.contains(end), self._gaps)
+        ):
+            # Make type system happy
+            ret = deepcopy(self[start_index:end_index])
+            if not isinstance(ret, float):
+                return ret
+            raise TypeError("Type error while copying data.")
```

Comment:
No, again, `deepcopy` is to return a copy of the requested buffer that can be modified without modifying the actual ringbuffer memory. No relation to types here (it existed before as well)

in vim I use pyright which complained about those lines a lot even though mypy and pylint didn't complain, but those errors also propagated into other parts and there mypy/pylint would complain sometimes, so I rather fixed it at the source.

`deepcopy` can return a float because if you pass a single-integer to `__getitem__` you get a single float value back, which in this case, is impossible as we pass a range. The typesystem doesn't know that though.

Reply:
> No, again, deepcopy is to return a copy of the requested buffer that can be modified without modifying the actual ringbuffer memory. No relation to types here (it existed before as well)

:+1:

> in vim I use pyright which complained about those lines a lot even though mypy and pylint didn't complain, but those errors also propagated into other parts and there mypy/pylint would complain sometimes, so I rather fixed it at the source.
>
> deepcopy can return a float because if you pass a single-integer to __getitem__ you get a single float value back, which in this case, is impossible as we pass a range. The typesystem doesn't know that though.

Ah, I see, then the correct way to type-hint this is to add `@overload`s for `__getitem__`, this tells the type system which combination of arguments and return types are valid. This way you won't have to do the runtime check yourself and the type system can make sure your types will be OK at runtime.

You can [see an example](https://github.com/leandro-lucarella-frequenz/frequenz-sdk-python/blob/55916591a2659690b88639fa05684e5d8b8136e3/src/frequenz/sdk/timeseries/_resampling/_buffer.py#L135-L143) in my failed attempt to make a more efficient ring buffer for the resampler:

```py
    @overload
    def __getitem__(self, index: int) -> T:
        ...


    @overload
    def __getitem__(self, index: slice) -> Sequence[T]:
        ...


    def __getitem__(self, index: int | slice) -> T | Sequence[T]:
```